xds/src/main/java/io/grpc/xds/AddressFilter.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.Attributes;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.NameResolver.ResolutionResultAttr;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+final class AddressFilter {
+  @ResolutionResultAttr
+  private static final Attributes.Key<PathChain> PATH_CHAIN_KEY =
+      Attributes.Key.create("io.grpc.xds.AddressFilter.PATH_CHAIN_KEY");
+
+  // Prevent instantiation.
+  private AddressFilter() {}
+
+  /**
+   * Sets a path filter to an EquivalentAddressGroup. Any preexisting path filter will be discarded.
+   */
+  static EquivalentAddressGroup setPathFilter(EquivalentAddressGroup address, List<String> names) {
+    checkNotNull(address, "address");
+    checkNotNull(names, "names");
+    Attributes.Builder attrBuilder = address.getAttributes().toBuilder().discard(PATH_CHAIN_KEY);
+    PathChain pathChain = null;
+    for (String name : names) {
+      if (pathChain == null) {
+        pathChain = new PathChain(name);
+        attrBuilder.set(PATH_CHAIN_KEY, pathChain);
+      } else {
+        pathChain.next = new PathChain(name);
+      }
+    }
+    return new EquivalentAddressGroup(address.getAddresses(), attrBuilder.build());
+  }
+
+  /**
+   * Selects addresses from the given list such that the first element in the path filter is
+   * identical to the given name, and returns the selected addresses with that first element removed
+   * from the path filter.
+   */
+  static List<EquivalentAddressGroup> filter(List<EquivalentAddressGroup> addresses, String name) {
+    checkNotNull(addresses, "addresses");
+    checkNotNull(name, "name");
+    List<EquivalentAddressGroup> filteredAddresses = new ArrayList<>();
+    for (EquivalentAddressGroup address : addresses) {
+      PathChain pathChain =
+          checkNotNull(address.getAttributes().get(PATH_CHAIN_KEY), "Not a hierarchical address");
```

Comment:
It may not be a good practice to make such a strong assumption for a utility method (it makes the method weak). You can just skip that address if the attribute does not exist.

Reply:
Skip the address now. I found C is doing the same.